Further attempts at building with .NET 8

Numpsy opened the pull request:
### Description

As https://github.com/fsprojects/FAKE/pull/2818, but updated on top of the latest changes and with a couple of extra commits to try to get more tests passing - see further comments for details

Thorium committed:
Build with .NET 8.0:

Change global.json .NET 6 to 8
Search and replace targetframeworks from fsproj files to add net8.0
Add net8.0 to paket.dependencies
dotnet paket install to find .NET8 compatible dependencies
Expecto had to be hardcoded for now, because some tests are running on netstandard2.0 library (hopefully we can update this separately later)
MSBuild.StructuredLogger problem: DisableInternalBinLog = true had to be added to build.fsx NuGet commands (hopefully we can update this separately later)
A few places of code had new overrides so had to explicitly type to strings
SdkAssemblyResolver to default .NET 8 as well
Readme update
GitHub pipeline configs: Add .NET 8 install.

Thorium committed:
try to run integration-tests with net6.0

Thorium committed:
uninstall didn't seem to work well anyway

Thorium committed:
publish both test-artifacts, ensure test DirectorySeparatorChar

Thorium committed:
parentheses added

Thorium committed:
integration tests back to 8

Thorium committed:
typo fix

Numpsy committed:
fix merge

Numpsy committed:
Try using the .NET 8 SDK for the resolver tests

Numpsy committed:
Try to work around https://github.com/dotnet/sdk/issues/40655

Numpsy commented on src/test/Fake.Core.IntegrationTests/SimpleHelloWorldTests.fs:
```diff
@@ -47,13 +47,19 @@ let tests =
               let scenarioPath = resolvePath scenario ""
               // dotnet tool install --version 5.19.0-alpha.local.1 fake-cli --add-source /e/Projects/FAKE/release/dotnetcore/
+
+              // Work around https://github.com/dotnet/sdk/issues/40655 by specifying the tool manifest explicitly
```
Some of tests in previous runs seemed to be failing with https://github.com/dotnet/sdk/issues/40655, whereby installing the dotnet tool in a subdirectory that contains a tools config file will ignore that file and use the one in the repo root, but then trying to use that tool *will* use the local config file and then fail.
That issue is marked as resolved, but it still seemed to be occuring in the SDK version used by the tests

Numpsy commented on build.fsx:
```diff
@@ -843,7 +840,7 @@ Target.create "CacheDotNetReleases" (fun _ ->
         |> Async.RunSynchronously
         |> List.ofSeq
-        |> List.find (fun product -> product.ProductVersion.Equals("6.0"))
+        |> List.find (fun product -> product.ProductVersion.Equals("8.0"))
```
I'm not sure what the logic for this (or cachedDotnetSdkReleases.json) should be when updating .NET versions. 
The current cachedDotnetSdkReleases.json only contains the .NET 6 releases and that seemed to cause test failures when built as .NET 8 and looking for the v8 SDK versions.
Changing this to 8.0 fixed that, but that might then not be right if fake-cli is multi-targetted at .NET 6 and .NET 8 as the .NET 6 version might still need the list of v6 releases?

Numpsy commented:
The remaining failures are in the TemplateIntegrationTests, with errors like
```
25h[11:48:10 ERR] Fake.DotNet.Cli.IntegrationTests.TemplateTests.can install and run the template.fails to build a target that doesn't exist failed in 00:00:09.6230000. 
Expected the message 'Target Nonexistent is not defined' but got: No exact match of product releases 8.0.401 found.
No product release found for 8.0.401. Maybe a pre-release? Returning all the versions.
No exact match of product releases 8.0.404 found.
No product release found for 8.0.404. Maybe a pre-release? Returning all the versions.
There was a problem while setting up the environment:
-> Could not find referenced assemblies in path: '[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.36\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.35\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.33\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.32\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.31\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.30\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.29\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.28\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.27\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.26\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.25\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.24\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.23\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.22\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.21\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.20\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.19\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.18\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.16\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.15\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.14\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.13\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.12\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.11\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.10\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.9\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.8\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.7\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.6\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.5\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.4\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.3\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.2\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.1\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-rc.2.21480.5\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-rc.1.21451.13\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.7.21377.19\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.6.21352.12\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.5.21301.5\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.4.21253.7\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.3.21201.4\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.2.21154.6\ref\net6.0],[C:\Users\runneradmin\AppData\Local\Microsoft\dotnet\packs\Microsoft.NETCore.App.Ref\6.0.0-preview.1.21102.12\ref\net6.0]', please check installed SDK and runtime versions
```
and I'm not sure if this is really a problem with the new stuff, or if it's because it's trying to use the 6.1.3 release to test the templates, and failing when running as .NET 8? (I suppose it shou;dn't be looking at the 6.0 assemblies when running as 8.0)

Thorium commented:
I expect that there is a chicken-and-egg problem, and the integration tests are actually restoring the .NET6 fake to try compiling .NET8, instead of referring to the just built .NET 8 version.

I tried this branch locally on my machine, and my observations are here:

`src\app\fake-cli\bin\Release\` has now `net6.0` and `net8.0` paths. Can a dotnet tool even have multiple target frameworks? Or should we just accept that FAKE 6.1.3 is working well for .NET6 projects and from now on everything else is .NET 8 only?

I run on my Windows machine dotnet build, and then from the actions:
` dotnet tool install fake-cli --add-source "./src/app/fake-cli/bin/Debug" --add-source "./src/app/fake-cli/bin/Release" --version 1.0.0-1`

It turned out that this didn't work properly. So I removed the Clean-task and called fake manually:

`C:\git\FAKE>src\app\fake-cli\bin\Debug\net8.0\fake-cli -v build -e FAKE_DETAILED_ERRORS=true`

And behold, all the integration tests went through, after painfully long half-an-hour later:

<img width="679" height="467" alt="image" src="https://github.com/user-attachments/assets/3d65a196-1740-44ca-a649-e1d0540c29c8" />


Where it did finally fail is:

```
dotnet.exe  deb --runtime linux-x64 --framework net8.0 --configuration Release --output "C:\git\FAKE\release\dotnetcore" --no-restore`

error NETSDK1045: The current .NET SDK does not support targeting .NET 8.0.  Either target .NET 6.0 or lower, or use a version of the .NET SDK that supports .NET 8.0. [C:\git\FAKE\src\app\fake-cl
```

So basically, this has to be updated from 0.1.220 to version 0.1.232

https://github.com/fsprojects/FAKE/blob/73fc9619df80596269621fcd607eff22102678cc/.config/dotnet-tools.json#L6

...and that's it, I think we could consider merging this and releasing a beta or new major version.

Numpsy commented:
> well for .NET6 projects and from now on everything else is .NET 8 only?

Whether there's any benefit to supporting .NET 6 in fake-cli any more is another question, and given the amount of pain in testing it all, just doing 8 seems simpler for a new major release?

.NET 10 is on the horizon, so making sure it works for that going forward seems more important than keeping 6 working in new versions. (My personal opinion anyway).

I'll see about rebasing it and fixing the conflicts anyway

Thorium commented:
Many projects use a separate non-FAKE fsproj as a build-project and reference the FAKE libraries, and there I see value in having .NET6 support kept. 

But the fake-cli tool, I think, could target only the NET8 (or even 10 later).

xperiandri commented:
Then it will be impossible to run it unless you have the latest SDK

Numpsy commented:
> Then it will be impossible to run it unless you have the latest SDK installed

The 'DNX' stuff that MS are talking about with .NET 10 (https://github.com/dotnet/core/blob/main/release-notes/10.0/preview/preview6/sdk.md#platform-specific-net-tools) could maybe help if the tool could be self contained, but that's a work in progress so it remains to be seen

Numpsy commented:
@Thorium I didn't see your other PR until after i'd rebased the changes to fix the merge conflicts, which caused conflicts in that, so I've cherry picked your dotnet-deb update straight into this PR

Thorium commented:
I think this file https://github.com/fsprojects/FAKE/blob/master/src/template/fake-template/Content/.config/dotnet-tools.json
says "use the nuget version of FAKE in the integration test templates", which causes any major changes to FAKE (like .NET update) to break the integration tests, because the NuGet version doesn't have these changes yet. 

But I'm not sure if there is any workaround possible.

Thorium commented:
One more thing: Because the system uses FSharp.Compiler.Service v43.8, the fsdocs-tool has to be upgraded from 16 to 20, to make GenerateDocs task pass.

Thorium committed:
Because the system uses FSharp.Compiler.Service v43.8, the fsdocs-tool has to be upgraded from 16 to 20, to make GenerateDocs task pass.

xperiandri commented:
@arialdomartini

Numpsy committed:
Fix merge

Numpsy commented:
General question about all the FAKE modules - should they all be both .NET 6 and .NET 8, or just NET 8? (or maybe only 6 if they don't have any dependencies where it matters?)

Numpsy commented on src/app/fake-cli/fake-cli.fsproj:
```diff
@@ -3,7 +3,7 @@
         <OutputType>Exe</OutputType>
         <ToolCommandName>fake</ToolCommandName>
-        <TargetFramework>net6.0</TargetFramework>
+        <TargetFrameworks>net8.0;net6.0</TargetFrameworks>
```
Maybe fake-cli itself should only be 8 regardless of what the modules are? (the integration tests are 8 only as it stands, so if it is still built as 6 as well it moght not get any testing)

Numpsy commented:
> I expect that there is a chicken-and-egg problem, and the integration tests are actually restoring the .NET6 fake to try compiling .NET8, instead of referring to the just built .NET 8 version.
> 
> I tried this branch locally (Numpsy/new8-build-bump-roll-tests-2) on my machine, and my observations are here:
> 
> `src\app\fake-cli\bin\Release\` has now `net6.0` and `net8.0` paths. Can a dotnet tool even have multiple target frameworks? Or should we just accept that FAKE 6.1.3 is working well for .NET6 projects and from now on everything else is .NET 8 only?
> 
> I run on my Windows machine dotnet build, and then from the actions: ` dotnet tool install fake-cli --add-source "./src/app/fake-cli/bin/Debug" --add-source "./src/app/fake-cli/bin/Release" --version 1.0.0-1`
> 
> It turned out that this didn't work properly. So I removed the Clean-task and called fake manually:
> 
> `C:\git\FAKE>src\app\fake-cli\bin\Debug\net8.0\fake-cli -v build -e FAKE_DETAILED_ERRORS=true`
> 
> And behold, all the integration tests went through, after painfully long half-an-hour later:
> <img alt="image" width="679" height="467" src="https://private-user-images.githubusercontent.com/229355/467476511-3d65a196-1740-44ca-a649-e1d0540c29c8.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjE1MjAxMjEsIm5iZiI6MTc2MTUxOTgyMSwicGF0aCI6Ii8yMjkzNTUvNDY3NDc2NTExLTNkNjVhMTk2LTE3NDAtNDRjYS1hNjQ5LWUxZDA1NDBjMjljOC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUxMDI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MTAyNlQyMzAzNDFaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT01MWUwYzg2OGRkMGY0YjM0OTk4MjJiZDM4MDIxZGQ0ZmIwMWFiNjBjOTg0MDAwYzUzY2I1ZDU1Nzg3M2M5NjM0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.Rv7AOC1MEfFJ89oqa05tusVCfmbYtw4P4SVO5XycteY">
> 
> Where it did finally fail is:
> 
> ```
> dotnet.exe  deb --runtime linux-x64 --framework net8.0 --configuration Release --output "C:\git\FAKE\release\dotnetcore" --no-restore`
> 
> error NETSDK1045: The current .NET SDK does not support targeting .NET 8.0.  Either target .NET 6.0 or lower, or use a version of the .NET SDK that supports .NET 8.0. [C:\git\FAKE\src\app\fake-cl
> ```
> 
> So basically, dotnet deb tool has to be updated from 0.1.220 to version 0.1.232
> 
> https://github.com/fsprojects/FAKE/blob/73fc9619df80596269621fcd607eff22102678cc/.config/dotnet-tools.json#L6
> 
> ...and that's it, I think we could consider merging this and releasing a beta or new major version.

It appears to work (all integration tests pass) if the .NET 6 runtime files are present in the local .NET install created by the tests, but not otherwise.

So

If i run the tests from this branch on a clean setup, if fails as the CI does.

If I run the build from the master branch, it installs the .NET 6 libs into ```%AppData%\Local\Microsoft\dotnet``` to run the .NET 6 tests, and leaves the files there.

If I then run the tests from this branch again, it installs the .NET 8 libs into ```%AppData%\Local\Microsoft\dotnet``` alongside the .NET 6 libs rather than replacing them, and the template integration tests find those and the tests pass.

Having .NET 6 installed in the default system location doesn't effect this as the tests are looking in the custom install location.

Numpsy committed:
Try to get the template integration tests working by making them use the 8.0.0 alpha build

Instead of pulling the 6.1.X build off NuGet.org.
This might bootstrap the tests enough to get them all passing

Numpsy commented on src/app/Fake.Runtime/SdkAssemblyResolver.fs:
I'm not sure if it should be both 6/8 now, or just 8 (will something running on .NET 8 ever want the .NET 6 libs?)
@Thorium ?

Numpsy committed:
update

Numpsy committed:
Change SdkAssemblyResolver.SdkVersions to take an empty string to mean 8.0.0 instead of 6.0.0

Numpsy committed:
In GetProductReleasesForSdk(), rename 'net60releases' to 'netReleases'

Because it might be .NET 8 releases now

Numpsy committed:
Update the Microsoft.Build libraries from 17.11.4 to 17.11.48 to pick up the fix for https://github.com/advisories/GHSA-w3q9-fxm7-j8fq

Also updates transitive dependencies on System.Text.Json to

Thorium commented:
Should we skip net8 and go straight to net10.0 ?

Numpsy commented:
Is there any value in completing #2880 and #2883 (on top on the unreleased #2879) for a last V6 release before moving everything to 8+, or is it not worth it?

xperiandri commented:
#2880 does not pass. But I would merge if it does.
I think it is a good idea to include in v6 minor release

Numpsy commented:
> #2880 does not pass. But I would merge if it does. I think it is a good idea to include in v6 minor release

The build failure there looks like it needs Fantomas running on it

Numpsy commented on RELEASE_NOTES.md:
```diff
@@ -1,4 +1,8 @@
 # Release Notes
+## 8.0.0-alpha - 2025-10-25
```
I suppose this date should be updated to this year before merging?

Numpsy committed:
Update release notes
